### DISCUSSION TITLE
Broadlink TC2 RF issue

### DISCUSSION THREAD

**Discussion by davecyf** [mlfunston/node-red-contrib-broadlink-control#65] [Ideas]
I try to integrate Broadlink TC2 with node-red .  As many of us failed, it is not working for TC2 RF.  I checked the message payload send to the RM unit, and compare it to what I get from rmplugin (a good android app) and find some interesting correlation.  The code send is actually the decimal value from 2's complement, and if you convert it back to Hex,  it is exactly what the RM unit need.  E.g. For the TC2 device, the first data byte is -23 (decimal value from 2's complement) which is E9 in Hex meaning 433Mhz in RM unit, so if the message payload is translated in the same way,  RM should able to fire the RF signal with Hex code.  I double-check with another old TC device, the first byte will translate to d7 in hex which is 315Mh in Broadlink RM unit.   That's only my guess.

**Comment by mlfunston** [comment]
It could definitely be something to look into. I don't have any units to test with though. Is there anyone else that wants to take a crack at testing and see?

**Reply by davecyf** [reply]
I supply more raw data for study:

The data array capture by a node from Broadlink e-control json files of a typical TC2 device: 

[0 … 9]
0: -23    (= 0xe9)
1: 20     (= 0x14)
2: 70     (= 0x46)
3: 0       ( is 0)
4: -33    (= 0xdf) 
5: 9       (= 0x09)
6: 22     (= 0x16)  and so on 
7: 9
8: 9
9: 22
........... The above data is decimal 2's complement of the below Hex version

Code translates from Android RM Plugin application,   import the same TC2 from the same json data of e-control 
Here is the coding scheme:

e9 = 433Mhz
14 = repeat: 20
46 00 = 70 double-bytes followed
Value Lo: 9
Value Hi: 22
Coding scheme: 2 Lo-Hi/Hi-Lo pairs
Only Lo/Hi values: false
Livolo code: false
000 •df -> 223
001 •09 -> 9 -> Lo
002 •16 -> 22 -> Hi
003 •09 -> 9 -> Lo
004 •09 -> 9 -> Lo
005 •16 -> 22 -> Hi
006 •16 -> 22 -> Hi
007 •09 -> 9 -> Lo
008 •09 -> 9 -> Lo
009 •16 -> 22 -> Hi
...........

I put here a complete TC2 hex code for reference

Hex:

e9144600df091609091616090916091616090916
160916090916091609160916160916090916091
616091609091609160916091616090916091616
091609091609160916160909000148

Hope that someone will take a look on it.
